crates/sui-move-build/src/linters/native_primitives_only.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This analysis flags uses of the sui::coin::Coin struct in fields of other structs. In most cases
+//! it's preferable to use sui::balance::Balance instead to save space.
+
+use move_command_line_common::{address::NumericalAddress, parser::NumberFormat};
+use move_compiler::{
+    diag,
+    diagnostics::codes::{custom, DiagnosticInfo, Severity},
+    expansion::ast as E,
+    naming::ast as N,
+    shared::{CompilationEnv, Identifier},
+    typing::{ast as T, core::ProgramInfo, visitor::TypingVisitor},
+};
+use move_core_types::account_address::AccountAddress;
+use move_ir_types::location::Loc;
+use move_symbol_pool::Symbol;
+
+use super::{
+    LinterDiagCategory, COIN_MOD_NAME, COIN_STRUCT_NAME, LINTER_DEFAULT_DIAG_CODE,
+    LINT_WARNING_PREFIX, SUI_PKG_NAME,
+};
+
+const NATIVE_PRIMITIVE_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::NativePrimitivesOnly as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "passing non-native types to native functions",
+);
+
+pub struct NativePrimitivesOnlyVisitor;
+
+impl TypingVisitor for NativePrimitivesOnlyVisitor {
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
+        for (_, _, mdef) in program.modules.iter() {
+            env.add_warning_filter_scope(mdef.warning_filter.clone());
+            mdef.functions
+                .iter()
```

Comment:
`
```suggestion
            mdef.functions
                .key_cloned_iter()
```
Using `key_cloned_iter` would be more in line with the rest of the compiler